index.coffee

```diff
@@ -148,20 +35,171 @@ onProgressHandler = (onProgressPromise, fallbackOnProgress) ->
 	# real onProgress function when the promise resolves
 	return (evt) -> onProgress(evt)
 
-getLongId = (shortId, layerIds) ->
-	if not shortId?
-		throw new Error('Progress event missing layer id. Progress not correct.')
-	longId = _.find(layerIds, (id) -> _.startsWith(id, shortId))
-	if not longId?
-		throw new Error("Progress error: Unknown layer #{shortId} referenced by docker. Progress not correct.")
-	return longId
+combinedProgressHandler = (renderer, states, index, callback) ->
+	return (evt) ->
+		# update current reporter state
+		states[index].percentage = evt.percentage
+		# update totals
+		percentage = _.sumBy(states, 'percentage') // (states.length || 1)
+		# update event
+		evt.totalProgress = renderer(percentage)
+		evt.percentage = percentage
+		evt.progressIndex = index
+		# call callback with aggregate event
+		callback(evt)
+
+class ProgressTracker
+	constructor: (@coalesceBelow = 0) ->
+		@layers = {}
+
+	addLayer: (id) ->
+		@layers[id] = { progress: null, coalesced: false }
+
+	updateLayer: (id, progress) ->
+		return if not id?
+		@addLayer(id) if not @layers[id]
+		@patchProgressEvent(progress)
+		@layers[id].coalesced = not @layers[id].progress? and progress.total < @coalesceBelow
+		@layers[id].progress = (progress.current / progress.total) || 0 # prevent NaN when .total = 0
```

Comment:
I'm just being careful I guess, .total is input coming from Docker which [sometimes](https://github.com/resin-io/docker-progress/pull/10#issuecomment-255407195) is unreliable.